src/vision/shiny_future/alan_switches_runtimes.md

```diff
@@ -0,0 +1,148 @@
+
+## The story
+
+Since his [early adventures](./alans_trust_in_the_compiler_is_rewarded.md) with Async I/O went so well, Alan has been looking for a way to learn more. He finds a job working in Rust. One of the projects he works on is [DistriData]. Looking at their code, he sees an annotation he has never seen before:
+
+```rust
+#[async_runtime(humboldt)]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let result = std::async_thread::spawn(async move {
+        do_something()
+    });
+}
+```
+
+He asks Barbara, one of his coworkers, "What is this `async_runtime` annotation? And what's `humboldt`?" She answers by explaining to him that Rust's support for async I/O is actually based around an underlying runtime. "Rust gives you a pretty decent runtime by default," she says, "but it's not tuned for our workloads. We wrote our own runtime, which we call `humboldt`."
+
+[DistriData]: ../projects/DistriData.md
+
+Alan asks, "What happens with the various std APIs? For example, I see we are calling `std::async_thread::spawn` -- when I used that before, it spawned tasks into the default runtime. What happens now?"
+
+Barbara explains that the "async" APIs in std generally execute relative to the current runtime that is in use. "When you call `std::async_thread::spawn`, it will spawn a task onto the current runtime. It's the same with the routines in `std::async_io` and and so forth. The `async_runtime` annotation just affects which runtime runs the `main` function, everything else follows from there."
+
+## Learning more about Humboldt
+
+Alan sees that some of the networking code that is being used in their application is creating network connections using humboldt APIs:
+
+```rust
+use humboldt::network;
+```
+
+He asks Barbara, "Why don't we use the `std::async_io` APIs for that?" She explains that Humbolt makes use of some custom kernel extensions that, naturally enough, aren't part of the std library. "TCP is for rubes," she says, "we are using TTCP -- Turbo TCP." Her mind wanders briefly to [Turbo Pascal] and she has a brief moment of yearning for the days when computers had a "Turbo" button that changed them from 8 MHz to 12 MHz. She snaps back into the present day. "Anyway, the `std::async_io` APIs just call into humboldt's APIs via various traits. But we can code directly against `humboldt` when we want to access the extra capabilities it offers."
+
+[Turbo Pascal]: https://en.wikipedia.org/wiki/Turbo_Pascal
+
+## Integrating into other event loops
+
+Later on, Alan is working on a visualizer front-end that integrates with [DistriData] to give more details about their workloads. To do it, he needs to integrate with Cocoa APIs and he wants to run certain tasks on Grand Central Dispatch. He approaches Barbara and asks, "If everything is running on `humboldt`, is there a way for me to run some things on another event loop? How does that work?"
```

Comment:
What are the Cocoa APIs and what exactly is Grand central Dispatch?

Reply:
From the following text I suspect that Grand Central Dispatch (gcd) is another runtime, is that right?

Reply:
[Grand Central Dispatch on wikipedia](https://en.wikipedia.org/wiki/Grand_Central_Dispatch) -- I don't really know whether it's the best example, I just wanted something that is specific to some setting.

Reply:
Ah okay. No it's totally fine to use that, but a short description what it is + link to Wikipedia would help to understand it.